targets/psa_switch/psa_switch.cpp

```diff
@@ -354,31 +372,42 @@ PsaSwitch::ingress_thread() {
     Deparser *deparser = this->get_deparser("ingress_deparser");
     deparser->deparse(packet.get());
 
-    // handling multicast
-    unsigned int mgid = 0u;
-    const auto &f_mgid = phv->get_field("psa_ingress_output_metadata.multicast_group");
-    mgid = f_mgid.get_uint();
-
+    // multicast
+    unsigned int mgid = phv->get_field("psa_ingress_output_metadata.multicast_group").get_uint();
     if(mgid != 0){
-      BMLOG_DEBUG_PKT(*packet,
-                      "Multicast requested for packet with multicast group {}",
-                      mgid);
-      const auto pre_out = pre->replicate({mgid});
-      auto packet_size = packet->get_register(PACKET_LENGTH_REG_IDX);
-      for(const auto &out : pre_out){
-        auto egress_port = out.egress_port;
-        BMLOG_DEBUG_PKT(*packet, "Replicating packet on port {}", egress_port);
-        std::unique_ptr<Packet> packet_copy = packet->clone_with_phv_ptr();
-        packet_copy->set_register(PACKET_LENGTH_REG_IDX, packet_size);
-        enqueue(egress_port, std::move(packet_copy));
-      }
+      multicast(packet.get(), mgid);
       continue;
     }
+    // TODO(pete) from psa spec - else if multicast group is empty, drop packet
+
+    // i2e clone
+    unsigned int clone = phv->get_field("psa_ingress_output_metadata.clone").get_uint();
+    unsigned int clone_session_id = phv->get_field("psa_ingress_output_metadata.clone_session_id").get_uint();
+    if (clone && clone_session_id) {
```

Comment:
I believe in PSA, 0 is a valid `clone_session_id`. This is why there is the `clone` boolean flag.

Reply:
Thanks, fixing this in next commit.